Upgrade react router to a stable release

ajainarayanan opened the pull request:
This PR is restarting efforts that went in as part of this PR https://github.com/caskdata/cdap/pull/8385

+ Add unit tests for components related to this change.

Components that require unit tests,

```
- EntityListView
- Overview (all entities overview)
- Detailed view (all entities detailed view)
- Header (including ProductDropdown, NamespaceDropdown)
- BreadCrumb,
- 404
```
This is to start with. Will add unit tests for other components in subsequent PR.

Bamboo build: https://builds.cask.co/browse/CDAP-DRC6419-1

tonybach commented:
Seems like there's a problem with the detail view of an application. Navigating directly to the URL doesn't work either. Here's the error:

![screen shot 2017-04-18 at 10 21 18 am](https://cloud.githubusercontent.com/assets/6516002/25143944/1cde1cc4-2421-11e7-992d-5c3cba7729bf.png)

tonybach commented on cdap-ui/app/cdap/api/__mocks__/app.js:
```diff
@@ -0,0 +1,59 @@
+};
+export {MyAppApi};
+
+// const list = () => {};
```
Do you want to keep these commented-out code?

tonybach commented on cdap-ui/app/cdap/api/__mocks__/app.js:
```diff
@@ -0,0 +1,59 @@
+MyAppApi.get = function() {
+  let subject = new Rx.Subject();
+  setTimeout(() => {
+    return this.__app;
```
`subject.onNext(this.__app)`?

tonybach commented on cdap-ui/app/cdap/api/__mocks__/dataset.js:
```diff
@@ -0,0 +1,29 @@
+  return subject;
+};
+
+export {MyDatasetApi};
```
Not sure why you're doing `export {}` in some places, `export default` in some others, as well as `module.exports = {}`.

ajainarayanan replied:
@tonybach the original modules some have a default export and some have named exports. If the mocks that we write have to replace them then they need to do the same thing. Making exports of api/*.js to be consistent can come in another PR as that will have more changes across the code.

tonybach commented on cdap-ui/app/cdap/components/EntityListView/NoEntitiesMessage/__tests__/NoEntitiesMessage.test.js:
```diff
@@ -27,6 +27,9 @@ let noEntitiesMessage;
 
 describe('NoEntitiesMessage Unit tests - ', () => {
+  beforeEach(() => {
+    filtersAreApplied = jest.fn();
```
I think this is unnecessary as we already have the function expression on line 27. Let me know if I'm missing something.

ajainarayanan commented:
@tonybach Updated.

tonybach commented on cdap-ui/app/cdap/api/__mocks__/app.js:
```diff
@@ -0,0 +1,43 @@
+MyAppApi.get = function() {
+  let subject = new Rx.Subject();
+  setTimeout(() => {
+    return subject.onNext(this.__app);
```
is it `return subject.onNext(this.__app)` or just ` subject.onNext(this.__app)`?

ajainarayanan replied:
Not sure if that would make any difference but will change though.

tonybach commented on cdap-ui/app/cdap/api/__mocks__/app.js:
```diff
@@ -0,0 +1,43 @@
+MyAppApi.batchStatus = function() {
+  let subject = new Rx.Subject();
+  setTimeout(() => {
+    return this.__batchStatusStats;
```
` subject.onNext(this.__batchStatusStats)` here too?

ajainarayanan committed:
Upgrade react router to a stable release

 - Adds react-router v4 to package.json + generates new dll files
 - Modifies Home and cdap/main.js file routes
 - Updates Link & NavLink in Header component
 - Modifies EntityListView + children + SearchStore based on route changes
 - Modifies Overview, App, Dataset & Stream Overview components based on route changes
 - Modifies AppDetailedView based on router changes
 - Modifies DatasetDetailedView based on router changes
 - Modifies StreamDetailedView based on router changes
 - Modifies references to Link and NavLink in 404, BreadCrumb, Market, MarketPlaceUsecaseEntity, NamespaceDropdown, RouteToNamespace & SpotlightSearch
 - Fixes api/* module references in ApplicationMetrics, ArtifactMetrics, DatasetMetrics & StreamMetrics for unit testing
 - Modifies DatsetStreamCards and Table based on router changes
 - Fixes Timer component to verify if its unmounted while/before starting the counter
 - Adds mock file for images and font files for unit testing
 - Modifies NoEntitiesMessage unit test based on default search text update in the component
 - Adds mock apis to be used in unit tests
 - Adds unit tests for components that are affected based on react router updates
 - Fixes FileBrowser, FilePath and DatapPrepConnections components with react router upgrade
 - Adds unit tests for FileBrowser component

elfenheart commented on cdap-ui/app/cdap/components/Overview/index.js:
```diff
@@ -258,5 +262,5 @@ export default class Overview extends Component {
 
 Overview.propTypes = {
-  onCloseAndRefresh: PropTypes.func
+  onCloseAndRefresh: PropTypes.func,
```
nit: extra comma not needed

tonybach commented on cdap-ui/app/cdap/components/Overview/StreamOverview/index.js:
```diff
@@ -168,5 +167,5 @@ StreamOverview.propTypes = {
   entity: PropTypes.object,
   onClose: PropTypes.func,
-  onCloseAndRefresh: PropTypes.func
+  onCloseAndRefresh: PropTypes.func,
```
here as well

elfenheart commented:
Minor codestyle comments, but otherwise 👍  LGTM. 

Everything seems to work as well, so let's get this merged!

tonybach commented:
LGTM 👍

bdmogal commented:
Let's be careful about such PRs in future. For now, since it is in the beginning of the release, and this has been worked on and tested for a while, I think it is ok to merge this.
